ldms/src/sampler/zfs_topvdevs/zfs_topvdevs.c

```diff
@@ -0,0 +1,440 @@
+ * Initialize the structure as schema and add them to the base schema.
+ * Also calculate the size of memory needed per schema and add it to the ldms
+ * schema list.
+ ****************************************************************************/
+
+static int initialize_ldms_structs()
+{
+        /*ldms_record_t zpool_def;  a pointer */
+        ldms_record_t   zpool_vdev_def;  /* a pointer */
+        int rc;
+
+        log_fn(LDMSD_LDEBUG, SAMP" initialize()\n");
+
+        /* Create the schema */
+        base_schema_new(sampler_base);
+        if (sampler_base->schema == NULL)
+            goto err1;
+
+        /* create the vdev record */
+        zpool_vdev_def  = ldms_record_from_template("zfs_topvdevs_stats",
+			                            zfs_topvdevs,
+						    vdev_metric_ids);
+        if (zpool_vdev_def == NULL)
+            goto err2;
+
+        zpool_vdev_heap_sz = ldms_record_heap_size_get(zpool_vdev_def);
+        rc = ldms_schema_record_add(sampler_base->schema, zpool_vdev_def);
+        if (rc < 0)
+            goto err3;
+
+	index_store.vdev_rec_idx = rc;
+	zpool_vdev_list_len.counttype  = TOPVDEV_COUNT;
+	zpool_vdev_list_len.vdev_count = 0;
+        rc = zpool_iter(g_zfs, get_vdevs_count, &zpool_vdev_list_len);
+	/* add error for iter in case here */
+        rc = ldms_schema_metric_list_add(sampler_base->schema,
+                                         "zpool_vdev_list",
+                                         NULL,
+                                         zpool_vdev_list_len.vdev_count  * zpool_vdev_heap_sz);
+        if (rc < 0)
+            goto err2;
+
+	index_store.vdev_list_idx = rc;
+
+        /* Create the metric set */
+        base_set_new(sampler_base);
+        if (sampler_base->set == NULL)
+                goto err2;
+
+        return 0;
+
+err3:
+        /* We only manually delete record template when it
+         * hasn't been added to the schema yet */
+        ldms_record_delete(zpool_vdev_def);
+err2:
+        base_schema_delete(sampler_base);
+err1:
+        log_fn(LDMSD_LERROR, SAMP" initialization failed\n");
+        return -1;
+}
+
+
+
+/*****************************************************************************
+ * WHAT:
+ * 1) Initialize the sampler base schema.
+ * 2) Initialize all structure and memory.
+ * 3) initialize the zfslib to sample the zpools stats.
+ * CALLER:
+ * ldms daemon. In error the plugin is aborted.
+ ****************************************************************************/
+
+
+static int config(struct ldmsd_plugin *self,
+                  struct attr_value_list *kwl, struct attr_value_list *avl)
+{
+        int rc = 0;
+
+        log_fn(LDMSD_LDEBUG, SAMP" config() called\n");
+
+        sampler_base = base_config(avl, SAMP, "zfs_topvdevs", log_fn);
+        if ((g_zfs = libzfs_init()) == NULL) {
+            rc = errno;
+            ldmsd_log(LDMSD_LERROR,
+                      SAMP" : Failed to initialize libzfs: %d\n", errno);
```

Comment:
```suggestion
                      SAMP" : Failed to initialize libzfs: %s\n", libzfs_error_init(rc));
```

Producers a more user friendly error and handles detecting if the kmod was loaded.